### ISSUE TITLE
Poller will throw ObjectDisposed exception if Socket.Dispose() called

### ISSUE THREAD

**Issue by jasells** [zeromq/netmq#834]
### Environment

    NetMQ Version:    
    Operating System: 
    .NET Version:

**Comment by jasells** [comment]
I created a fork to repro by adding a new unit test [NetMQPollerTest.RemoveAndDisposeSocket()](https://github.com/jasells/netmq/blob/3b8dbb929f76046852c5eb88b5869f0fed3fdbe5/src/NetMQ.Tests/NetMQPollerTest.cs#L464).

This test as-s will never complete, as the exception kills the test execution, and there is no way to catch it to handle it gracefully from test code (or app code).  

Debug the test to see the exception.

**Comment by jasells** [comment]
I made branch [here ](https://github.com/jasells/netmq/blob/taskFix/src/NetMQ.Tests/NetMQPollerTest.cs)with a potential fix.

I refactored ```NetMQPoller.Remove(ISocketPollable)``` to return a ```Task``` instead of ```void```.

This allows for proper thread syncing of the async ops needed to properly remove a socket from ```Poller`` internally.

If .NET4.0 support could be removed, the fix can be made entirely internal and the change would be a non-breaking change to any code using it.

dropping .NET4.0 support would allow the method to be defined as ```async void Remove(ISocketPollable)``` and the ```await``` can be applied internally, maintaining the outward API as a syncronous-API.

Maybe there is some other workaround for .NET40?

**Comment by toonetown** [comment]
Yes - a million times YES!  This issue has been hitting us for years...we just haven't had much success in tracking it down.  Is there a corresponding pull request?

**Comment by jasells** [comment]
Will get one in tomorrow.  I have a better fix that doesn’t break API. Sent from Mail for Windows 10 From: Nathan TooneSent: Friday, December 6, 2019 10:22 AMTo: zeromq/netmqCc: jasells; AuthorSubject: Re: [zeromq/netmq] Poller will throw ObjectDisposed exception if Socket.Dispose() called (#834) Yes - a million times YES! This issue has been hitting us for years...we just haven't had much success in tracking it down. Is there a corresponding pull request?—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by jasells** [comment]
So, there is a PR, but it is a little old, and needs updates to resolve conflicts with master.

I have been side-tracked and responses were slow so I lost track when it was fresh in my mind.  I will try to get this PR resolved.
